core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java

```diff
@@ -71,51 +74,72 @@
+   * socket server
+   * 2. Short-Circuit - if the data source is the local worker
+   * 3. Local Loopback Read - if the data source is the local worker and short circuit is disabled
+   * 4. Read from remote worker - if the data source is a remote worker
+   * 5. UFS Read from worker - if the data source is UFS, read from the UFS policy's designated
+   * worker (ufs -> local or remote worker -> client)
    *
    * @param context the file system context
-   * @param blockId the block ID
-   * @param blockSize the block size in bytes
-   * @param address the Alluxio worker address
-   * @param blockSource the source location of the block
-   * @param openUfsBlockOptions the options to open a UFS block, set to null if this is block is
-   *        not persisted in UFS
-   * @param options the in stream options
-   * @return the {@link InputStream} object
+   * @param info the block info
+   * @param dataSource the Alluxio worker which should read the data
+   * @param dataSourceType the source location of the block
+   * @param options the instream options
+   * @return the {@link BlockInStream} object
    */
-  public static BlockInStream create(FileSystemContext context, long blockId, long blockSize,
-      WorkerNetAddress address, BlockInStreamSource blockSource,
-      Protocol.OpenUfsBlockOptions openUfsBlockOptions, InStreamOptions options)
-          throws IOException {
-    if (Configuration.getBoolean(PropertyKey.USER_SHORT_CIRCUIT_ENABLED)
-        && !NettyUtils.isDomainSocketSupported(address)
-        && blockSource == BlockInStreamSource.LOCAL) {
+  public static BlockInStream create(FileSystemContext context, BlockInfo info,
+      WorkerNetAddress dataSource, BlockInStreamSource dataSourceType, InStreamOptions options)
+      throws IOException {
+    URIStatus status = options.getStatus();
+    OpenFileOptions readOptions = options.getOptions();
+
+    boolean promote = readOptions.getReadType().isPromote();
+
+    long blockId = info.getBlockId();
+    long blockSize = info.getLength();
+
+    // Construct the partial read request
+    Protocol.ReadRequest.Builder builder =
+        Protocol.ReadRequest.newBuilder().setBlockId(blockId).setPromote(promote);
+    if (status.isPersisted()) { // Add UFS fallback options
+      builder.setOpenUfsBlockOptions(options.getOpenUfsBlockOptions(blockId));
+    }
+
+    boolean shortCircuit = Configuration.getBoolean(PropertyKey.USER_SHORT_CIRCUIT_ENABLED);
+    boolean sourceSupportsDomainSocket = !NettyUtils.isDomainSocketSupported(dataSource);
+    boolean sourceIsLocal = dataSourceType == BlockInStreamSource.LOCAL;
+
+    // Short circuit
+    if (sourceIsLocal && shortCircuit && !sourceSupportsDomainSocket) {
+      LOG.debug("Creating short circuit input stream for block {} @ {}", blockId, dataSource);
       try {
-        LOG.debug("Creating short circuit input stream for block {} @ {}", blockId, address);
```

Comment:
should we still keep this debug msg?

Reply:
The message is one line above outside the try block

Reply:
ack